<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Promise对象</title>
</head>

<body>
    <script>
        // const promise = new Promise(function (resolve, reject) {
        //     // 一些代码

        //     if () {
        //         resolve()
        //     } else {
        //         reject()
        //     }
        // })
        // promise.then(function(value) {
        //     // success
        // }, function(error) {
        //     // failure
        // })

        // function timeout(ms) {
        //     return new Promise((resolve, reject) => {
        //         setTimeout(resolve, ms, "done")
        //     })
        // }

        // timeout(100).then(value => {
        //     console.log(value)
        // })
        // console.log(1)
        // setTimeout(() => {
        //     console.log(2)
        // }, 90)

        // let promise = new Promise(function (resolve, reject) {
        //     console.log("Promise")
        //     resolve()
        // })

        // promise.then(function () {
        //     console.log("resolved")
        // })

        // console.log("hi!")
        // setTimeout(() => {
        //     console.log("setTimeout")
        // }, 0)
        /* 
        Promise
        hi!
        resolved
        */

        // function loadImageAsync(url) {
        //     return new Promise(function (resolve, reject) {
        //         const image = new Image()
        //         image.onload = function () {
        //             resolve(image)
        //         }
        //         image.onerror = function () {
        //             reject(new Error('Could not load image at ' + url))
        //         }
        //         image.src = url
        //     })
        // }

        // const getJson = function (url) {
        //     const promise = new Promise(function (resolve, reject) {
        //         const handler = function () {
        //             if (this.readState !== 4) {
        //                 return
        //             }
        //             if (this.status === 200) {
        //                 resolve(this.response)
        //             } else {
        //                 reject(new Error(this.statusText))
        //             }
        //         }
        //         const client = new XMLHttpRequest()
        //         client.open("GET", url)
        //         client.onreadstatechange = handler
        //         client.responseTpe = "json"
        //         client.setRequestHeader("Access", "application/json")
        //         client.send()
        //     })
        //     return promise
        // }
        // getJson("/post.json").then(function (json) {
        //     console.log("Content: " + json)
        // }, function (error) {
        //     console.error("出错了", error)
        // })

        // const p1 = new Promise(function (resolve, reject) {
        //     // some code
        // })

        // const p2 = new Promise(function (resolve, reject) {
        //     // ...
        //     resolve(p1)
        // })

        // const p1 = new Promise(function (resolve, reject) {
        //     setTimeout(() => {
        //         reject(new Error("fail"))
        //     }, 3000)
        // })

        // const p2 = new Promise(function (resolve, reject) {
        //     setTimeout(() => {
        //         resolve(p1)
        //     }, 1000)
        // })

        // p2.then(result => {
        //     console.log(result)
        // }).catch(error => {
        //     console.log("error", error) // error Error: fail
        // })
        // new Promise((resolve, reject) => {
        //     resolve(1);
        //     console.log(2);
        // }).then(r => {
        //     console.log(r);
        // });
        // new Promise((resolve, reject) => {
        //     return resolve(1)
        //     // 后面的语句不会执行
        //     console.log(2)
        // })
        // getJson("/post.json").then(function (json) {
        //     return json.post
        // }).then(function (post) {
        //     // ...
        // })

        // getJson("/post.json").then(function (post) {
        //     return getJson(post.commentURL)
        // }).then(function (comments) {
        //     console.log("resolved", comments)
        // }, function (error) {
        //     console.log("rejected", error)
        // })

        // getJson("/post.json").then(post =>
        //     getJson(post.commentURL)
        // ).then(
        //     comments =>
        //     console.log("resolved", comments),
        //     error =>
        //     console.log("rejected", error)
        // )

        // Promise.proptotype.cathc
        // getJson("/post.json").then(function (post) {
        //     // 
        // }).cathc(function (error) {
        //     // 处理getJso 和前一个回调函数运行发生的错误
        //     console.log("发生错误！", error)
        // })

        // p.then(val => {
        //     console.log("fulfilled", val)
        // }).cathc(error => {
        //     console.log("rejected", error)
        // })

        // // 等同于
        // p.then(val => {
        //     console.log("fulfilled", val)
        // }).then(null, error => {
        //     console.log("rejected", error)
        // })

        // const promise = new Promise(function (resolve, reject) {
        //     throw new Error("test")
        // })
        // promise.catch(function (error) {
        //     console.log(error) // Error: test
        // })

        // 写法一
        // const promise = new Promise(function (resolve, reject) {
        //     try {
        //         throw new Error("test")
        //     } catch (e) {
        //         reject(e)
        //     }
        // })
        // promise.catch(error => {
        //     console.log(error) // Error: test
        // })

        // // 写法二
        // const promise = new Promise(function (resolve, reject) {
        //     reject(new Error("test"))
        // })
        // promise.catch(error => {
        //     console.log(error) // Error: test
        // })

        // const promise = new Promise(function (resolve, reject) {
        //     resolve("ok")
        //     throw new Error("test")
        // })
        // promise.then(value => {
        //     console.log(value) // ok
        // }).catch(error => {
        //     console.log(error)
        // })

        // const someAsyncThing = function () {
        //     return new Promise(function (resolve, reject) {
        //         // 下面一行会报错，因为x，没有定义
        //         return (x + 2)
        //     })
        // }
        // someAsyncThing().then(() => {
        //     console.log("everything is great")
        // })

        // setTimeout(() => {
        //     console.log(123)
        // }, 2000)
        // // Promise对象.html:222 Uncaught (in promise) ReferenceError: x is not defined
        // // 123

        // const promise = new Promise(function (resolve, reject) {
        //     resolve("ok")
        //     setTimeout(() => {
        //         throw new Error("test")
        //     }, 0)
        // })
        // promise.then(value => {
        //     console.log(value)
        // })
        // // ok
        // // Uncaught Error: test

        // const someAsyncThing = function () {
        //     return new Promise(function (resolve, reject) {
        //         // 下面一行会报错，因此x没有声明
        //         resolve(x + 2)
        //     })
        // }
        // someAsyncThing().catch(error => {
        //     console.log("oh on", error) // oh on ReferenceError: x is not defined
        // }).then(() => {
        //     console.log("carry on") // carry on
        // })

        // Promise.resolve().catch(error => {
        //     console.log("oh on", error)
        // }).then(() => {
        //     console.log("carry on") // carry on
        // })

        // const someAsyncThing = function () {
        //     return new Promise(function (resolve, reject) {
        //         // 下面一行会报错，因此x没有声明
        //         resolve(x + 2)
        //     })
        // }
        // someAsyncThing().then(() => {
        //     return someAsyncThing()
        // }).catch(error => {
        //     console.log("oh on", error) // oh on ReferenceError: x is not defined
        //     // 下面这一行会报错，以为y没有声明
        //     y + 2 // Uncaught (in promise) ReferenceError: y is not defined
        // }).then(() => {
        //     console.log("carry on")
        // })
        // someAsyncThing().then(() => {
        //     return someAsyncThing()
        // }).catch(error => {
        //     console.log("oh on", error) // oh on ReferenceError: x is not defined
        //     // 下面这一行会报错，以为y没有声明
        //     y + 2
        // }).catch(error => {
        //     console.log("carry on", error) // carry on ReferenceError: y is not defined
        // })

        // const promise = new Promise(function (resolve, reject) {})
        // promise.then(result => {
        //     // ...
        // }).catch(error => {
        //     // ...
        // }).finally(() => {
        //     // ...
        // })

        // server.listen(port).then(() => {
        //     // ...
        // }).finally(server.stop)

        // Promise.prototype.finally = function (callback) {
        //     let P = this.constructor
        //     return this.then(
        //         value => {
        //             P.resolve(callback()).then(() => value)
        //         },
        //         reason => P.resolve(callback()).then(() => {
        //             throw reason
        //         })
        //     )
        // }

        // // resolve 的值是 undefined
        // Promise.resolve(2).then(() => {}, () => {})

        // // resolve 的值是 2
        // Promise.resolve(2).finally(() => {})

        // // reject 的值是 undefined
        // Promise.reject(3).then(() => {}, () => {})

        // // reject 的值是 3
        // Promise.reject(3).finally(() => {})

        // Promise.all
        // const p = new Promise.all([p1, p2, p3])

        // // 生成一个Promise的对象数组
        // const promises = [2, 3, 5, 7, 11, 13].map(function (id) {
        //     return getJson("/post/" + id + ".json")
        // })
        // Promise.all(promises).then(function (posts) {
        //     // ...
        // }).catch(function (reason) {
        //     // ...
        // })

        // const databasePromise = connectDatabase()
        // const booksPromise = databasePromise.then(findAllBooks)
        // const userPromise = databasePromise.then(getCurrentUser)

        // Promise.all([booksPromise, userPromise]).then(([books, user]) => {
        //     pickTopRecommendations(books, user)
        // })

        // const p1 = new Promise((resolve, reject) => {
        //         resolve("hello")
        //     })
        //     .then(result => result)
        //     .catch(e => e)

        // const p2 = new Promise((resolve, reject) => {
        //         throw new Error("报错了")
        //     })
        //     .then(result => result)
        // .catch(e => e)

        // Promise.all([p1, p2])
        //     .then(result => console.log(result))
        //     .catch(e => console.log(e))
        // // ['hello', Error: 报错了]

        // const p1 = new Promise((resolve, reject) => {
        //         resolve("hello")
        //     })
        //     .then(result => result)

        // const p2 = new Promise((resolve, reject) => {
        //         throw new Error("报错了")
        //     })
        //     .then(result => result)

        // Promise.all([p1, p2])
        //     .then(result => console.log(result))
        //     .catch(e => console.log(e)) // Error: 报错了

        // Promise.race()
        // const p = new Promise.race([p1, p2, p3])

        // const p = Promise.race([fetch("/resource-that-my-take-a-while"),
        //     new Promise(function (resolve, reject) {
        //         setTimeout(() => reject(new Error("request timeout")), 5000)
        //     })
        // ])

        // p.then(console.log()).catch(console.error())

        // const promises = [
        //     fetch('/api-1'),
        //     fetch('/api-2'),
        //     fetch('/api-3'),
        // ];

        // await Promise.allSettled(promises);
        // removeLoadingIndicator();

        // const resolved = Promise.resolve(42)
        // const rejected = Promise.reject(-1)

        // const allSettledPromise = Promise.allSettled([resolved, rejected])

        // allSettledPromise.then(function (results) {
        //     console.log(results)
        // })
        // /* 
        // [
        // { status: 'fulfilled', value: 42 },
        // { status: 'rejected', reason: -1 }
        // ]
        // */

        // const promises = [fetch('index.html'), fetch('https://does-not-exist/')];
        // const results = await Promise.allSettled(promises);

        // // 过滤出成功的请求
        // const successfulPromises = results.filter(p => p.status === "fulfilled")

        // // 过滤出失败的请求
        // const errors = results
        //     .filter(p => p.status === "rejected")
        //     .map(p => p.reason)

        // const urls = [ /* ... */ ]
        // const requests = urls.map(x => fetch(x))

        // try {
        //     await Promise.all(requests)
        //     console.log("所有请求都成功！")
        // } catch {
        //     console.log("至少一个请求失败，其他请求可能还没有结束。")
        // }
        // const promises = [
        //     fetch('/endpoint-a').then(() => 'a'),
        //     fetch('/endpoint-b').then(() => 'b'),
        //     fetch('/endpoint-c').then(() => 'c'),
        // ];
        // try {
        //     const first = await Promise.any(promises);
        //     console.log(first);
        // } catch (error) {
        //     console.log(error);
        // }

        // new AggregateError() extends Array -> AggregateError
        // const err = new AggregateError();
        // err.push(new Error("first error"));
        // err.push(new Error("second error"));
        // throw err;

        // const resolved = Promise.resolve(42);
        // const rejected = Promise.reject(-1);
        // const alsoRejected = Promise.reject(Infinity);
        // Promise.any([resolved, rejected, alsoRejected]).then(function (result) {
        //     console.log(result) // 42
        // })

        // Promise.any([rejected, alsoRejected]).catch(function (results) {
        //     console.log(results) // AggregateError: All promises were rejected
        // })

        // const jsPromise = Promise.resolve($.ajax('/whatever.json'));

        // Promise.resolve('foo')
        // // 等价于
        // new Promise(resolve => resolve('foo'))

        // let thenable = {
        //     then: function (resolve, reject) {
        //         resolve(42);
        //     }
        // };

        // let p1 = Promise.resolve(thenable);
        // p1.then(function (value) {
        //     console.log(value); // 42
        // });

        // const p = Promise.resolve('hello')
        // p.then(function (s) {
        //     console.log(s) // hello
        // })
        // setTimeout(function () {
        //     console.log('five');
        // }, 0);

        // const p = new Promise(function (resolve, reject) {
        //     resolve("three")
        //     console.log("one")
        // })

        // p.then((value) => {
        //     console.log(value)
        // })
        // Promise.resolve().then(function () {
        //     console.log('four');
        // });

        // console.log('two');
        // /* 
        // // one
        // // two
        // // three
        // // four
        // // five
        // */

        // const p = Promise.reject('出错了');
        // // 等同于
        // const p = new Promise((resolve, reject) => reject('出错了'))

        // p.then(null, function (s) {
        //     console.log(s)
        // });

        // const thenable = {
        //     then(resolve, reject) {
        //         reject('出错了');
        //     }
        // };

        // Promise.reject(thenable)
        //     .catch(e => {
        //         console.log(e === thenable)
        //     })

        // const preloadImage = function (path) {
        //     return new Promise(function (resolve, reject) {
        //         const image = new Image()
        //         image.onload = resolve
        //         image.onerror = reject
        //         iamge.src = path
        //     })
        // }

        // Promise.resolve().then(f)

        // const f = () => console.log("now")

        // Promise.resolve().then(f)
        // console.log("next")
        // // next
        // // now

        // const f = () => console.log("now")

        // (async () => f())();
        // console.log("next")

        // const f = () => console.log('now');
        // (
        //     () => new Promise(
        //         resolve => resolve(f())
        //     )
        // )();
        // console.log('next');

        const f = () => console.log('now');
        Promise.try(f);
        console.log('next');
    </script>
</body>

</html>